### PR TITLE
feat(restdsl): Add openapi-restdsl generator REST endpoint

### LINKED ISSUE #709: Add openapi-restdsl generator REST endpoint
Generate exact same REST DSL camel-k and camel-jbang would generate when they consume OpenAPI spec, and use it for Kaoto REST DSL support. Add such REST endpoint at backend so that kaoto-ui/rest can request it.

### PR DESCRIPTION
**Pull request by igarashitm**
Fixes: #709

 
#### Open API spec in JSON
```
[tomo@p1 resources]$ curl -v -d @petstore.json -H "Content-Type: application/json" http://localhost:8081/v1/rest/openApiToRest
*   Trying 127.0.0.1:8081...
* Connected to localhost (127.0.0.1) port 8081 (#0)
> POST /rest/openapiToRest HTTP/1.1
> Host: localhost:8081
> User-Agent: curl/8.0.1
> Accept: */*
> Content-Type: application/json
> Content-Length: 3841
> 
< HTTP/1.1 200 OK
< content-length: 711
< Content-Type: text/yaml;charset=UTF-8
< 
- rest:
    path: "/v1"
    get:
    - id: "listPets"
      path: "/pets"
      produces: "application/json"
      param:
      - dataType: "integer"
        description: "How many items to return at one time (max 100)"
        name: "limit"
        required: false
        type: "query"
      to: "direct:listPets"
    - id: "showPetById"
      path: "/pets/{petId}"
      produces: "application/json"
      param:
      - dataType: "string"
        description: "The id of the pet to retrieve"
        name: "petId"
        required: true
        type: "path"
      to: "direct:showPetById"
    post:
    - id: "createPets"
      path: "/pets"
      produces: "application/json"
      to: "direct:createPets"
* Connection #0 to host localhost left intact
```

#### Open API spec in YAML
```
[tomo@p1 resources]$ curl -v --data-binary @petstore.yaml -H "Content-Type: text/yaml" http://localhost:8081/v1/rest/openApiToRest?inputFormat=yaml 
*   Trying 127.0.0.1:8081...
* Connected to localhost (127.0.0.1) port 8081 (#0)
> POST /rest/openapiToRest?inputFormat=yaml HTTP/1.1
> Host: localhost:8081
> User-Agent: curl/8.0.1
> Accept: */*
> Content-Type: text/yaml
> Content-Length: 2618
> 
< HTTP/1.1 200 OK
< content-length: 711
< Content-Type: text/yaml;charset=UTF-8
< 
- rest:
    path: "/v1"
    get:
    - id: "listPets"
      path: "/pets"
      produces: "application/json"
      param:
      - dataType: "integer"
        description: "How many items to return at one time (max 100)"
        name: "limit"
        required: false
        type: "query"
      to: "direct:listPets"
    - id: "showPetById"
      path: "/pets/{petId}"
      produces: "application/json"
      param:
      - dataType: "string"
        description: "The id of the pet to retrieve"
        name: "petId"
        required: true
        type: "path"
      to: "direct:showPetById"
    post:
    - id: "createPets"
      path: "/pets"
      produces: "application/json"
      to: "direct:createPets"
* Connection #0 to host localhost left intact
```

### CODE REVIEW

**Comment by igarashitm** [comment]
#### Bonus - Generate XML REST DSL
```
[tomo@p1 resources]$ curl -v --data-binary @petstore.yaml -H "Content-Type: text/yaml" http://localhost:8081/rest/openapiToRest?inputFormat=yaml\&outputFormat=xml 
*   Trying 127.0.0.1:8081...
* Connected to localhost (127.0.0.1) port 8081 (#0)
> POST /rest/openapiToRest?inputFormat=yaml&outputFormat=xml HTTP/1.1
> Host: localhost:8081
> User-Agent: curl/8.0.1
> Accept: */*
> Content-Type: text/yaml
> Content-Length: 2618
> 
< HTTP/1.1 200 OK
< content-length: 810
< Content-Type: text/yaml;charset=UTF-8
< 
<?xml version="1.0" encoding="UTF-8" standalone="no"?><rests xmlns="http://camel.apache.org/schema/spring">
    <rest path="/v1">
        <get id="listPets" path="/pets" produces="application/json">
            <param dataType="integer" description="How many items to return at one time (max 100)" name="limit" required="false" type="query"/>
            <to uri="direct:listPets"/>
        </get>
        <post id="createPets" path="/pets" produces="application/json">
            <to uri="direct:createPets"/>
        </post>
        <get id="showPetById" path="/pets/{petId}" produces="application/json">
            <param dataType="string" description="The id of the pet to retrieve" name="petId" required="true" type="path"/>
            <to uri="direct:showPetById"/>
        </get>
    </rest>
* Connection #0 to host localhost left intact
```

**Comment by igarashitm** [comment]
It might be even possible to directly invoke `IntegrationsResource#integration()` and get the Kaoto JSON, if it's possible to inject `IntegrationResource` instance.

File: api-rest/src/main/java/io/kaoto/backend/api/rest/resource/OpenApiToRestDslResource.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+            final @Parameter(description = "Camel Rest DSL format, either YAML or XML. "
+                    + "Assumes YAML if not specified.", example = "xml")
+            @QueryParam("outputFormat") String outputFormat
+            ) throws Exception {
+        CamelContext context = new DefaultCamelContext();
+        var jacksonNode = "yaml".equalsIgnoreCase(inputFormat) ? readFromYaml(spec) : readFromJson(spec);
+        OasDocument specDoc = (OasDocument) Library.readDocument(jacksonNode);
+        return "xml".equalsIgnoreCase(outputFormat)
+                ? RestDslGenerator.toXml(specDoc).generate(context)
+                : RestDslGenerator.toYaml(specDoc).generate(context);
```
Comment (igarashitm) [inline api-rest/src/main/java/io/kaoto/backend/api/rest/resource/OpenApiToRestDslResource.java:69]: The main part is this few lines

**Comment by lordrip** [comment]
Open questions:

1. Could the roundtrip be achieved as well?
2. What would happen if I have a rest section and send it to Kaoto will this work fine as well?
3. For the input: could we remove this parameter and inspect this value at the backend side? the reason why I'm asking it's to support files as well, like uploading a local JSON to be parsed.
4. For the output: could we have a single word parameter like `output` and if not provided, by default it's YAML?

**Comment by igarashitm** [comment]
We'll discuss 1. and 2. more since it involves frontend workflow.

> 3. For the input: could we remove this parameter and inspect this value at the backend side? the reason why I'm asking it's to support files as well, like uploading a local JSON to be parsed.

Quick search suggests that `The only fool-proof way to know if something is valid YAML and not valid JSON, is to try to parse it as both. `
https://www.google.com/search?q=test+if+json+or+yaml&oq=test+if+json+or+yaml&aqs=chrome..69i57j69i64.3293j0j7&sourceid=chrome&ie=UTF-8

 We can do that parse-and-see-if-fail for sure, but if we were to add file/external URL support, I think it's gonna be a quick way like just check the extension. We can do that when we support URL. Or do we need it already?

BTW the `inputFormat` is assumend `JSON` if not specified.
https://github.com/KaotoIO/kaoto-backend/pull/710/files#diff-f7c72520204d03d2c60431e35934262c60aafeda5bda465f5092a9a7d4649aa6R65

> 4. For the output: could we have a single word parameter like `output` and if not provided, by default it's YAML?

Yes `output` sounds fine and it already assumes YAML if not specified
https://github.com/KaotoIO/kaoto-backend/pull/710/files#diff-f7c72520204d03d2c60431e35934262c60aafeda5bda465f5092a9a7d4649aa6R67

**Comment by igarashitm** [comment]
> I think it's gonna be a quick way like just check the extension

Maybe not. I'll add that fool-proof way anyway, then it also handles the case URL has a wrong extension.

**Comment by igarashitm** [comment]
#### The interaction between kaoto-ui-rest
In order to avoid circlular dependency or polluting `api` module with DSL dependent code, it's better to separate into 2 actions
1. Post Open API specification and generate Camel Rest DSL in YAML
2. Generate Kaoto JSON (Integration) object from YAML

When user requests to import Open API spec, following is the expected workflow
1. kaoto-ui-rest posts the Open API spec to this `openApiToRest` endpoint and receive a `rest` portion of Camel YAML DSL
2. kaoto-ui-rest then posts previously received `rest` portion of Camel YAML DSL to the existing YAML>Kaoto-JSON endpoint, and receive a `rest` portion of Kaoto-JSON (Integration/FlowsWrapper)
3. kaoto-ui-rest then merges previously received `rest` portion of Kaoto-JSON into existing `FlowsWrapper` model object

Since kaoto-backend is stateless, it should be safe to do 2) YAML>Kaoto-JSON request separately from other routes/rests/etc.

**Comment by igarashitm** [comment]
JVM tests passed locally, stuck on native tests...

**Review by Delawen — CHANGES_REQUESTED** [review]

File: api-rest/src/main/java/io/kaoto/backend/api/rest/resource/v1/OpenApiToRestDslResource.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            final @Parameter(description = "Camel Rest DSL format, either YAML or XML. "
+                    + "It assumes YAML if not specified.", example = "xml")
+            @QueryParam("output") String output
+            ) throws Exception {
+        var runtime = Arc.container().instance(CamelRuntime.class).get();
+        var camelContext = runtime.getCamelContext();
+        var jacksonNode = readOpenApiSpec(spec);
+        OasDocument specDoc = (OasDocument) Library.readDocument(jacksonNode);
+        return "xml".equalsIgnoreCase(output)
+                ? RestDslGenerator.toXml(specDoc).generate(camelContext)
```
[the remark this answers is not in the document]
Comment (igarashitm) [reply api-rest/src/main/java/io/kaoto/backend/api/rest/resource/v1/OpenApiToRestDslResource.java]: removed :+1:

**Review by Delawen — APPROVED** [review]

**Pull request merged**